MagazineLayout/LayoutCore/ModelState.swift

```diff
@@ -730,6 +736,146 @@ final class ModelState {
+    in rect: CGRect,
+    withElementLocationsForFlattenedIndices elementLocationsForFlattenedIndices: [Int: ElementLocation],
+    andFramesProvidedBy frameProvider: ((ElementLocation) -> CGRect))
+    -> ElementLocationFramePairs
+  {
+    var elementLocationFramePairs = ElementLocationFramePairs()
+
+    guard
+      let indexOfFirstElement = indexOfFirstElement(
+        in: rect,
+        withElementLocationsForFlattenedIndices: elementLocationsForFlattenedIndices,
+        andFramesProvidedBy: frameProvider) else
+    {
+      return elementLocationFramePairs
+    }
+
+    // Used to handle the case where we encounter an element that doesn't intersect the rect, but
+    // previous elements in the same row might.
+    var minYOfNonIntersectingElement: CGFloat?
+
+    // Look backward to find visible elements
+    for elementLocationIndex in (0..<indexOfFirstElement).reversed() {
+      let elementLocation = self.elementLocation(
+        forFlattenedIndex: elementLocationIndex,
+        in: elementLocationsForFlattenedIndices)
+      let frame = frameProvider(elementLocation)
+
+      guard frame.maxY > rect.minY else {
+        if let minY = minYOfNonIntersectingElement, frame.minY < minY {
+          // We're in a previous row, so we know we've captured all intersecting rects for the
+          // subsequent row.
+          break
+        } else {
+          // We've found a non-intersecting item, but still need to check other items in the same
+          // row.
+          minYOfNonIntersectingElement = frame.minY
+          continue
+        }
+      }
+
+      elementLocationFramePairs.append(
+        ElementLocationFramePair(elementLocation: elementLocation, frame: frame))
+    }
+
+    // Look forward to find visible elements
+    for elementLocationIndex in indexOfFirstElement..<elementLocationsForFlattenedIndices.count {
+      let elementLocation = self.elementLocation(
+        forFlattenedIndex: elementLocationIndex,
+        in: elementLocationsForFlattenedIndices)
+      let frame = frameProvider(elementLocation)
+      guard frame.minY < rect.maxY else { break }
+
+      elementLocationFramePairs.append(
+        ElementLocationFramePair(elementLocation: elementLocation, frame: frame))
+    }
+
+    return elementLocationFramePairs
+  }
+
+  private func indexOfFirstElement(
+    in rect: CGRect,
+    withElementLocationsForFlattenedIndices elementLocationsForFlattenedIndices: [Int: ElementLocation],
+    andFramesProvidedBy frameProvider: ((ElementLocation) -> CGRect))
+    -> Int?
+  {
+    var lowerBound = 0
+    var upperBound = elementLocationsForFlattenedIndices.count - 1
+
+    while lowerBound <= upperBound {
+      let index = (lowerBound + upperBound) / 2
+      let elementLocation = self.elementLocation(
+        forFlattenedIndex: index,
+        in: elementLocationsForFlattenedIndices)
+      let elementFrame = frameProvider(elementLocation)
+      if elementFrame.maxY <= rect.minY {
```

Comment:
We could use a switch here

Reply:
I don't think so actually, since the first comparison is checking `elementFrame.maxY`, and the second comparison is checking `elementFrame.minY`